narayana-spring-boot-core/src/main/java/me/snowdrop/boot/narayana/core/jdbc/lrco/LrcoXAConnection.java

```diff
@@ -0,0 +1,398 @@
+import java.sql.Struct;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Executor;
+
+import javax.sql.ConnectionEvent;
+import javax.sql.ConnectionEventListener;
+import javax.sql.StatementEventListener;
+import javax.sql.XAConnection;
+import javax.transaction.xa.XAResource;
+
+public class LrcoXAConnection implements XAConnection {
+
+    private final Connection physicalConnection;
+    private volatile Connection handleConnection;
+    private final List<ConnectionEventListener> eventListeners;
+
+    public LrcoXAConnection(Connection connection) throws SQLException {
+        this.physicalConnection = connection;
+        this.eventListeners = new ArrayList<>();
+    }
+
+    @Override
+    public XAResource getXAResource() throws SQLException {
+        return new LrcoXAResource(this.physicalConnection);
+    }
+
+    @Override
+    public Connection getConnection() throws SQLException {
+        Connection lastHandleConnection = this.handleConnection;
+        if (lastHandleConnection != null) {
+            lastHandleConnection.close();
+        }
+        this.physicalConnection.rollback();
+        this.handleConnection = new PooledJdbcConnection();
+        return this.handleConnection;
+    }
+
+    @Override
+    public void close() throws SQLException {
+        Connection lastHandleConnection = this.handleConnection;
+        if (lastHandleConnection != null) {
+            lastHandleConnection.close();
+        }
+        this.physicalConnection.close();
+    }
+
+    private void closeHandle() {
+        ConnectionEvent event = new ConnectionEvent(this);
+        for (int i = 0; i < this.eventListeners.size(); i++) {
+            ConnectionEventListener listener = this.eventListeners.get(i);
+            listener.connectionClosed(event);
+        }
+        this.handleConnection = null;
+    }
+
+    @Override
+    public void addConnectionEventListener(ConnectionEventListener listener) {
+        if (!this.eventListeners.contains(listener)) {
+            this.eventListeners.add(listener);
+        }
+    }
+
+    @Override
+    public void removeConnectionEventListener(ConnectionEventListener listener) {
+        this.eventListeners.remove(listener);
+    }
+
+    @Override
+    public void addStatementEventListener(StatementEventListener listener) {
+    }
+
+    @Override
+    public void removeStatementEventListener(StatementEventListener listener) {
+    }
+
+    private final class PooledJdbcConnection implements Connection {
+
+        private boolean closed;
+
+        private PooledJdbcConnection() {
+            this.closed = false;
+        }
+
+        @Override
+        public void close() throws SQLException {
+            if (!this.closed) {
+                try {
+                    LrcoXAConnection.this.physicalConnection.rollback();
+                    LrcoXAConnection.this.physicalConnection.setAutoCommit(true);
```

Comment:
Why do you need to set autocomit to true here?

Reply:
This is something most lrco impls I read are doing after reset original connection to set autocommit back to true. Some even saves old content.